File: Cpp/697_Degree_of_an_Array.cpp
class Solution {
public:
    int findShortestSubArray(vector<int>& nums) {
        map<int, int>freq;
        map<int, int>start;
        int max_degree = 1;
        int min_length = nums.size();
        for(int i = 0; i < nums.size(); ++i){
            if(freq[nums[i]] == 0){
                start[nums[i]] = i;
            }
            freq[nums[i]] +=1;
            if (freq[nums[i]] > max_degree) {
                max_degree = freq[nums[i]];
                min_length = i-start[nums[i]] + 1;
            } else if (freq[nums[i]] == max_degree) {
                min_length = min(min_length, i-start[nums[i]]+1);
            }
        }
        return min_length;
    }
};

File: Kotlin/697_Degree_of_an_Array.kt
class Solution {
    fun findShortestSubArray(nums: IntArray): Int {
        var frequency = hashMapOf<Int, Int>()
        var start = hashMapOf<Int, Int>()
        var max_degree = 1
        var length = nums.lastIndex + 1

        for((i, num) in nums.withIndex()) {
            if(!frequency.containsKey(nums[i])){
                start[num] = i
            }
            frequency[num] = (frequency[num]?:0) + 1
            if(frequency[num]?:0 > max_degree) {
                max_degree = frequency[num]?:0
                length = i - (start[num]?:0) + 1
            }else if (frequency[num] == max_degree) {
                length = Math.min(length, i - (start[num]?:0) + 1)
            }
        }
        
        return length
    }
}